Tests/WorkspaceTests/WorkspaceTests.swift

```diff
@@ -660,6 +660,67 @@ final class WorkspaceTests: XCTestCase {
         }
     }
 
+    func testEditAndPinning() throws {
+        let path = AbsolutePath("/RootPkg")
+        let fs = InMemoryFileSystem()
+        let manifestGraph = try MockManifestGraph(at: path,
+            rootDeps: [
+                MockDependency("A", version: Version(1, 0, 0)..<Version(1, .max, .max)),
+                MockDependency("B", version: Version(1, 0, 0)..<Version(1, .max, .max)),
+            ],
+            packages: [
+                MockPackage("A", version: v1),
+                MockPackage("B", version: v1),
+            ],
+            fs: fs
+        )
+        let provider = manifestGraph.repoProvider!
+
+        // Create the workspace.
+        let workspace = Workspace.createWith(rootPackage: path,
+                                             manifestLoader: manifestGraph.manifestLoader,
+                                             delegate: TestWorkspaceDelegate(),
+                                             fileSystem: fs,
+                                             repositoryProvider: provider)
+        // Load the package graph.
+        let engine = DiagnosticsEngine()
+        let graph = workspace.loadPackageGraph(rootPackages: [path], engine: engine)
+        XCTAssertFalse(engine.hasErrors())
+        let rootManifests = workspace.loadRootManifests(packages: [path], engine: engine)
+        XCTAssertFalse(engine.hasErrors())
+        let manifests = workspace.loadDependencyManifests(rootManifests: rootManifests, engine: engine)
+        XCTAssertFalse(engine.hasErrors())
+        guard let aManifest = manifests.lookup(manifest: "A") else {
+            return XCTFail("Expected manifest for package A not found")
+        }
+
+        func getDependency(_ manifest: Manifest) -> ManagedDependency {
```

Comment:
Off topic: We should build some more testing infra for Workspace, a lot of similar code is repeated in lots of tests unfortunately.